File: test/integration/test_csv_tguf_roundtrip.cpp
#include <gtest/gtest.h>

#include <cstdint>
#include <cstdlib>
#include <string>

#include "tgn.h"

class CSV_TGUF_RoundtripTest : public ::testing::Test {
 protected:
  const std::string script_path = "scripts/convert_csv_to_tguf.sh";
  const std::string resource_dir =
      "test/integration/resources/csv_tguf_roundtrip/";
  const std::string edges_csv = resource_dir + "edges.csv";
  const std::string labels_csv = resource_dir + "labels.csv";
  const std::string node_feats_csv = resource_dir + "node_feats.csv";
  const std::string output_tguf = resource_dir + "out.tguf";
  const std::string cmd = std::format("{} {} {} {} {}", script_path, edges_csv,
                                      output_tguf, labels_csv, node_feats_csv);

  void SetUp() override { ASSERT_EQ(std::system(cmd.c_str()), 0); }
};

TEST_F(CSV_TGUF_RoundtripTest, Verify) {
  auto store = tgn::TGStore::from_tguf(output_tguf);

  // TODO(kuba): Would be nice to read in expect out from resources
  EXPECT_EQ(store->edge_count(), 3);
  EXPECT_EQ(store->node_count(), 31);
  EXPECT_EQ(store->msg_dim(), 2);
  EXPECT_EQ(store->label_dim(), 2);
  EXPECT_EQ(store->node_feat_dim(), 3);

  // Check Edges
  auto batch = store->get_batch(0, 3, tgn::TGStore::NegStrategy::PreComputed);

  EXPECT_EQ(batch.src[0].item<std::int64_t>(), 1);
  EXPECT_EQ(batch.src[1].item<std::int64_t>(), 2);
  EXPECT_EQ(batch.src[2].item<std::int64_t>(), 3);

  EXPECT_EQ(batch.dst[0].item<std::int64_t>(), 20);
  EXPECT_EQ(batch.dst[1].item<std::int64_t>(), 30);
  EXPECT_EQ(batch.dst[2].item<std::int64_t>(), 10);

  EXPECT_EQ(batch.time[0].item<std::int64_t>(), 5);
  EXPECT_EQ(batch.time[1].item<std::int64_t>(), 10);
  EXPECT_EQ(batch.time[2].item<std::int64_t>(), 15);

  EXPECT_FLOAT_EQ(batch.msg[0][0].item<float>(), 0.11F);
  EXPECT_FLOAT_EQ(batch.msg[2][0].item<float>(), 0.31F);

  // Check Pre-Computed Negatives
  EXPECT_EQ(batch.neg_dst.value()[0][0].item<std::int64_t>(), 9);
  EXPECT_EQ(batch.neg_dst.value()[0][1].item<std::int64_t>(), 8);
  EXPECT_EQ(batch.neg_dst.value()[1][0].item<std::int64_t>(), 7);
  EXPECT_EQ(batch.neg_dst.value()[1][1].item<std::int64_t>(), 6);
  EXPECT_EQ(batch.neg_dst.value()[2][0].item<std::int64_t>(), 5);
  EXPECT_EQ(batch.neg_dst.value()[2][1].item<std::int64_t>(), 4);

  // Check labels
  auto label0 = store->get_label_event(0);
  EXPECT_EQ(label0.n_id[0].item<std::int64_t>(), 1);
  EXPECT_FLOAT_EQ(label0.target[0][0].item<float>(), 1.0F);
  EXPECT_FLOAT_EQ(label0.target[0][1].item<float>(), 0.0F);

  auto label1 = store->get_label_event(1);
  EXPECT_EQ(label1.n_id[0].item<std::int64_t>(), 2);
  EXPECT_FLOAT_EQ(label1.target[0][0].item<float>(), 0.0F);
  EXPECT_FLOAT_EQ(label1.target[0][1].item<float>(), 1.0F);

  // Check Edge-Label Synchronization (Stop IDs)
  // Label 0 (t=12) sees edges before 12 (Index 0, 1) -> Stop ID = 2
  EXPECT_EQ(store->get_edge_cutoff_for_label_event(0), 2);
  // Label 1 (t=24) sees all edges (Index 0, 1, 2) -> Stop ID = 3
  EXPECT_EQ(store->get_edge_cutoff_for_label_event(1), 3);

  // Check node features
  const auto n_id = torch::arange(
      static_cast<std::int64_t>(store->node_count()), torch::kLong);
  const auto node_feats = store->gather_node_feats(n_id);
  EXPECT_EQ(node_feats.size(0), n_id.size(0));
  EXPECT_EQ(node_feats.size(1), store->node_feat_dim());
  for (auto i = 0; i < node_feats.size(0); ++i) {
    if (i % 5 == 0 && i <= 20) {
      // Values in CSV are (node_id / 5) * multiplier
      const auto multiplier = static_cast<float>(i) / 5.0F;
      EXPECT_EQ(node_feats[i][0].item<float>(), 1.0F * multiplier);
      EXPECT_EQ(node_feats[i][1].item<float>(), 2.0F * multiplier);
      EXPECT_EQ(node_feats[i][2].item<float>(), 3.0F * multiplier);
    } else {
      EXPECT_EQ(node_feats[i][0].item<float>(), 0.0F);
      EXPECT_EQ(node_feats[i][1].item<float>(), 0.0F);
      EXPECT_EQ(node_feats[i][2].item<float>(), 0.0F);
    }
  }
}

File: python/test/test_csv_tguf_roundtrip.py
import csv
from pathlib import Path

import numpy as np
import pytest
import torch

import tguf


@pytest.fixture
def resource_dir():
    return Path(__file__).parent / "resources" / "csv_tguf_roundtrip"


@pytest.fixture
def output_tguf(tmp_path):
    return tmp_path / "out.tguf"


def load_csv_data(path: Path):
    with open(path, mode="r", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        headers = reader.fieldnames
        rows = list(reader)
    return headers, rows


def test_csv_tguf_roundtrip(resource_dir, output_tguf):
    e_headers, e_rows = load_csv_data(resource_dir / "edges.csv")
    l_headers, l_rows = load_csv_data(resource_dir / "labels.csv")
    n_headers, n_rows = load_csv_data(resource_dir / "node_feats.csv")

    msg_cols = [c for c in e_headers if c.startswith("msg_")]
    neg_cols = [c for c in e_headers if c.startswith("neg_")]
    label_cols = [c for c in l_headers if c.startswith("node_y")]
    feat_cols = [c for c in n_headers if c.startswith("node_x")]

    all_ids = []
    for r in e_rows:
        all_ids.extend([int(r["src"]), int(r["dst"])])
    for r in l_rows:
        all_ids.append(int(r["node_id"]))
    for r in n_rows:
        all_ids.append(int(r["node_id"]))
    max_id = max(all_ids)

    schema = tguf.TGUFSchema(
        path=str(output_tguf),
        edge_capacity=len(e_rows),
        msg_dim=len(msg_cols),
        label_dim=len(label_cols),
        node_feat_capacity=int(max_id + 1),
        node_feat_dim=len(feat_cols),
        label_capacity=len(l_rows),
        negatives_per_edge=len(neg_cols),
    )
    builder = tguf.TGUFBuilder(schema)

    builder.append_edges(
        tguf.Batch(
            src=np.array([int(r["src"]) for r in e_rows], dtype=np.int64),
            dst=np.array([int(r["dst"]) for r in e_rows], dtype=np.int64),
            time=np.array([int(r["time"]) for r in e_rows], dtype=np.int64),
            msg=np.array(
                [[float(r[c]) for c in msg_cols] for r in e_rows], dtype=np.float32
            ),
            neg_dst=np.array(
                [[int(r[c]) for c in neg_cols] for r in e_rows], dtype=np.int64
            )
            if neg_cols
            else None,
        )
    )

    builder.append_labels(
        n_id=np.array([int(r["node_id"]) for r in l_rows], dtype=np.int64),
        time=np.array([int(r["time"]) for r in l_rows], dtype=np.int64),
        target=np.array(
            [[float(r[c]) for c in label_cols] for r in l_rows], dtype=np.float32
        ),
    )

    builder.append_node_feats(
        n_id=np.array([int(r["node_id"]) for r in n_rows], dtype=np.int64),
        node_feat=np.array(
            [[float(r[c]) for c in feat_cols] for r in n_rows], dtype=np.float32
        ),
    )

    builder.finalize()

    store = tguf.TGStore.from_tguf(str(output_tguf))

    assert store.edge_count == 3
    assert store.node_count == 31
    assert store.msg_dim == 2
    assert store.label_dim == 2
    assert store.node_feat_dim == 3

    batch = store.get_batch(0, 3, strategy=tguf.NegStrategy.PreComputed)
    torch.testing.assert_close(batch.src, torch.tensor([1, 2, 3], dtype=torch.int64))
    torch.testing.assert_close(batch.dst, torch.tensor([20, 30, 10], dtype=torch.int64))
    torch.testing.assert_close(batch.time, torch.tensor([5, 10, 15], dtype=torch.int64))

    expected_negs = torch.tensor([[9, 8], [7, 6], [5, 4]], dtype=torch.int64)
    torch.testing.assert_close(batch.neg_dst, expected_negs)

    label0 = store.get_label_event(0)
    assert label0.n_id[0] == 1
    torch.testing.assert_close(
        label0.target[0], torch.tensor([1.0, 0.0], dtype=torch.float32)
    )

    assert store.get_edge_cutoff_for_label_event(0) == 2
    assert store.get_edge_cutoff_for_label_event(1) == 3

    n_ids = torch.arange(store.node_count, dtype=torch.int64)
    node_feats = store.gather_node_feats(n_ids)

    for i in range(len(n_ids)):
        if i % 5 == 0 and i <= 20:
            mult = i / 5.0
            expected = torch.tensor(
                [1.0 * mult, 2.0 * mult, 3.0 * mult], dtype=torch.float32
            )
            torch.testing.assert_close(node_feats[i], expected)
        else:
            torch.testing.assert_close(node_feats[i], torch.zeros(3))
